server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -81,6 +102,102 @@ func (b *BuildEventHandler) OpenChannel(ctx context.Context, iid string) interfa
 		hasReceivedStartedEvent: false,
 		eventsBeforeStarted:     make([]*inpb.InvocationEvent, 0),
 		logWriter:               logWriter,
+		onClose:                 onClose,
+	}
+}
+
+func (b *BuildEventHandler) GetStatsRecorder() *statsRecorder {
+	return b.statsRecorder
+}
+
+// recordStatsRequest is a request to record the stats for an invocation, which
+// will be fulfilled by one of the workers spawned by StatsRecorder.
+type recordStatsRequest struct {
+	invocationID string
+	requestedAt  time.Time
+}
+
+// statsRecorder listens for finalized invocations and copies cache stats from
+// the metrics collector to the DB.
+type statsRecorder struct {
+	env          environment.Env
+	openChannels *sync.WaitGroup
+	workerDone   chan struct{}
+
+	mu      sync.Mutex // protects(stopped, reqs)
+	reqs    chan *recordStatsRequest
+	stopped bool
+}
+
+func newStatsRecorder(env environment.Env, openChannels *sync.WaitGroup) *statsRecorder {
+	return &statsRecorder{
+		env:          env,
+		openChannels: openChannels,
+		reqs:         make(chan *recordStatsRequest, 4096),
+		workerDone:   make(chan struct{}),
+	}
+}
+
+func (r *statsRecorder) MarkFinalized(invocationID string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.stopped {
+		alert.UnexpectedEvent(
+			"stats_recorder_finalize_after_shutdown",
+			"Invocation %q was marked finalized after the stats recorder was shut down.",
+			invocationID)
+	}
+
+	req := &recordStatsRequest{
+		invocationID: invocationID,
+		requestedAt:  time.Now(),
+	}
+	r.reqs <- req
+}
+
+func (r *statsRecorder) Start() {
+	for i := 0; i < numStatsRecorderWorkers; i++ {
+		r.startWorker(context.Background())
+	}
+}
+
+func (r *statsRecorder) startWorker(ctx context.Context) {
+	go func() {
+		defer func() {
+			r.workerDone <- struct{}{}
+		}()
+
+		for req := range r.reqs {
+			// Apply the finalization delay relative to when the invocation was marked
+			// finalized, rather than relative to now. Otherwise each worker would be
+			// unnecessarily throttled.
+			time.Sleep(time.Until(req.requestedAt.Add(cacheStatsFinalizationDelay)))
+			ti := &tables.Invocation{InvocationID: req.invocationID}
+			if stats := hit_tracker.CollectCacheStats(ctx, r.env, req.invocationID); stats != nil {
+				fillInvocationFromCacheStats(stats, ti)
+			}
+			if err := r.env.GetInvocationDB().InsertOrUpdateInvocation(ctx, ti); err != nil {
+				log.Errorf("Failed to write cache stats for invocation: %s", err)
+			}
+		}
+	}()
+}
+
+func (r *statsRecorder) Stop() {
+	// Wait for all EventHandler channels to be closed to ensure there will be
+	// no more calls to MarkFinalized.
+	r.openChannels.Wait()
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.stopped = true
+	close(r.reqs)
+
+	// Wait for all workers to be done recording stats.
+	for i := 0; i < numStatsRecorderWorkers; i++ {
+		<-r.workerDone
 	}
 }
```

Comment:
I think you can simplify a fair bit of this code by using an errorgroup instead.

func (r *statsRecorder) Start() {
  r.eg := errgroup.Group{}
	for i := 0; i < numStatsRecorderWorkers; i++ {
		eg.Go(func() error {
            for req := range r.reqs {
			// Apply the finalization delay relative to when the invocation was marked
			// finalized, rather than relative to now. Otherwise each worker would be
			// unnecessarily throttled.
			time.Sleep(time.Until(req.requestedAt.Add(cacheStatsFinalizationDelay)))
			ti := &tables.Invocation{InvocationID: req.invocationID}
			if stats := hit_tracker.CollectCacheStats(ctx, r.env, req.invocationID); stats != nil {
				fillInvocationFromCacheStats(stats, ti)
			}
			if err := r.env.GetInvocationDB().InsertOrUpdateInvocation(ctx, ti); err != nil {
				log.Errorf("Failed to write cache stats for invocation: %s", err)
			}
		}
        })
	}
}

func (r *statsRecorder) Stop() {
	// Wait for all EventHandler channels to be closed to ensure there will be
	// no more calls to MarkFinalized.
	r.openChannels.Wait()

	r.mu.Lock()
	defer r.mu.Unlock()

	r.stopped = true
	close(r.reqs)
    r.eg.Wait()
}

Reply:
Nice - done!